From 3ed8102c8a881e6fe273bbf06f45e86dccf51e02 Mon Sep 17 00:00:00 2001
From: Evgeny Vereshchagin <evvers@ya.ru>
Date: Sun, 4 Oct 2026 07:16:01 +0000
Subject: [PATCH] core: set the receive options before binding the sockets

It fixes a bug where avahi crashed with
```
assertion "found_addr" failed: file "socket.c", line 794, function "avahi_recv_dns_packet_ipv4"

Program received signal SIGABRT, Aborted.
0x00007f7ff78b7f7a in _lwp_kill () from /usr/lib/libc.so.12
(gdb) bt
 #0  0x00007f7ff78b7f7a in _lwp_kill () from /usr/lib/libc.so.12
 #1  0x00007f7ff78c7584 in abort () from /usr/lib/libc.so.12
 #2  0x00007f7ff787ea2d in __assert13 () from /usr/lib/libc.so.12
 #3  0x00007f7ff7d4f2c1 in avahi_recv_dns_packet_ipv4 (fd=12, ret_src_address=0x7f7fffffdf24, ret_src_port=0x7f7fffffdf70, ret_dst_address=0x7f7fffffdee4,
    ret_iface=0x7f7fffffdf60, ret_ttl=0x7f7fffffdf80 "\377\n.\367\337{") at socket.c:794
 #4  0x00007f7ff7d13341 in mcast_socket_event (w=0x7bdff72e0980, fd=12, events=AVAHI_WATCH_IN, userdata=0x7cdff72e2d80) at server.c:1081
 #5  0x00007f7ff7ec1302 in avahi_simple_poll_dispatch (s=0x7c5ff72e0040) at simple-watch.c:585
 #6  0x00007f7ff7ec192b in avahi_simple_poll_iterate (s=0x7c5ff72e0040, timeout=-1) at simple-watch.c:605
 #7  0x00000000005a8cd5 in run_server (c=0xff5300 <config>) at main.c:1279
 #8  0x000000000059aadf in main (argc=1, argv=0x7f7fffffe708) at main.c:1708
```
and in the case of IPv6 with
```
assertion "found_iface" failed: file "socket.c", line 916, function "avahi_recv_dns_packet_ipv6"
...
 #3  0x00007f7ff7d50d91 in avahi_recv_dns_packet_ipv6 (fd=13, ret_src_address=0x7f7fffffdf24, ret_src_port=0x7f7fffffdf70, ret_dst_address=0x7f7fffffdee4,
    ret_iface=0x7f7fffffdf60, ret_ttl=0x7f7fffffdf80 "X\n.\367\337{") at socket.c:916
 #4  0x00007f7ff7d1358d in mcast_socket_event (w=0x7bdff72e09e0, fd=13, events=AVAHI_WATCH_IN, userdata=0x7cdff72e2d80) at server.c:1085
 #5  0x00007f7ff7ec1302 in avahi_simple_poll_dispatch (s=0x7c5ff72e0040) at simple-watch.c:585
 #6  0x00007f7ff7ec192b in avahi_simple_poll_iterate (s=0x7c5ff72e0040, timeout=-1) at simple-watch.c:605
 #7  0x00000000005a8cd5 in run_server (c=0xff5300 <config>) at main.c:1279
 #8  0x000000000059aadf in main (argc=1, argv=0x7f7fffffe708) at main.c:1708
```
on NetBSD and OpenBSD on receiving packets queued after the sockets were
bound but before setsockopt was called. On Linux avahi didn't crash
but those packets came with ifindex=0 and were discarded later.

The bug was introduced back in 2005 and surfaced on the upstream NetBSD
CI where cname-test sporadically crashed.

Co-Authored-By: Steven Schlansker <stevenschlansker@gmail.com>
---
 avahi-core/socket.c | 25 ++++++++++++-------------
 1 file changed, 12 insertions(+), 13 deletions(-)

diff --git a/avahi-core/socket.c b/avahi-core/socket.c
index 302fb5b..c755834 100644
--- a/avahi-core/socket.c
+++ b/avahi-core/socket.c
@@ -336,6 +336,9 @@ int avahi_open_socket_ipv4(int no_reuse) {
         goto fail;
     }
 
+    if (ipv4_pktinfo(fd) < 0)
+        goto fail;
+
     memset(&local, 0, sizeof(local));
     local.sin_family = AF_INET;
     local.sin_port = htons(AVAHI_MDNS_PORT);
@@ -348,9 +351,6 @@ int avahi_open_socket_ipv4(int no_reuse) {
     if (r < 0)
         goto fail;
 
-    if (ipv4_pktinfo (fd) < 0)
-         goto fail;
-
     if (avahi_set_cloexec(fd) < 0) {
         avahi_log_warn("FD_CLOEXEC failed: %s", strerror(errno));
         goto fail;
@@ -406,6 +406,9 @@ int avahi_open_socket_ipv6(int no_reuse) {
         goto fail;
     }
 
+    if (ipv6_pktinfo(fd) < 0)
+        goto fail;
+
     memset(&local, 0, sizeof(local));
     local.sin6_family = AF_INET6;
     local.sin6_port = htons(AVAHI_MDNS_PORT);
@@ -418,9 +421,6 @@ int avahi_open_socket_ipv6(int no_reuse) {
     if (r < 0)
         goto fail;
 
-    if (ipv6_pktinfo(fd) < 0)
-        goto fail;
-
     if (avahi_set_cloexec(fd) < 0) {
         avahi_log_warn("FD_CLOEXEC failed: %s", strerror(errno));
         goto fail;
@@ -933,6 +933,9 @@ int avahi_open_unicast_socket_ipv4(void) {
         goto fail;
     }
 
+    if (ipv4_pktinfo(fd) < 0)
+        goto fail;
+
     memset(&local, 0, sizeof(local));
     local.sin_family = AF_INET;
 
@@ -941,10 +944,6 @@ int avahi_open_unicast_socket_ipv4(void) {
         goto fail;
     }
 
-    if (ipv4_pktinfo(fd) < 0) {
-         goto fail;
-    }
-
     if (avahi_set_cloexec(fd) < 0) {
         avahi_log_warn("FD_CLOEXEC failed: %s", strerror(errno));
         goto fail;
@@ -979,6 +978,9 @@ int avahi_open_unicast_socket_ipv6(void) {
         goto fail;
     }
 
+    if (ipv6_pktinfo(fd) < 0)
+        goto fail;
+
     memset(&local, 0, sizeof(local));
     local.sin6_family = AF_INET6;
 
@@ -987,9 +989,6 @@ int avahi_open_unicast_socket_ipv6(void) {
         goto fail;
     }
 
-    if (ipv6_pktinfo(fd) < 0)
-        goto fail;
-
     if (avahi_set_cloexec(fd) < 0) {
         avahi_log_warn("FD_CLOEXEC failed: %s", strerror(errno));
         goto fail;
-- 
2.53.0

